### ISSUE TITLE
Filter operator for arrays

### ISSUE THREAD

**Issue by michaelhkay** [qt4cg/qtspecs#1159]
I propose to provide `?[...]` as a filter operator for arrays.

For example, 

```
let $array := [(1,2,3), (4,5,6,7)]
return $array?[count(.) = 4]
```
returns

`[(4,5,6,7)]`

I propose that the operator should work exactly like the familiar `[]` for sequences in its handling of numeric and boolean predicate values. So for example $array?[2,1] in the above example returns `[(4,5,6,7), (1,2,3)]`. The result is always an array (which may be a little surprising). This means that `$array?[3]` has the same effect as `[$array?3]` or `[$array(3)]`.

**Comment by michaelhkay** [comment]
We could also use this syntax for filtering maps, with the convention that "." is bound to a KVP representing the map entry. So, for example

```
let $map := {"x1":1, "y":200, "x3":50}
return $map?[?key => starts-with('x') and ?value > 3]
```

returns `{"x3":50}`

**Comment by michaelhkay** [comment]
Closed by acceptance of PR #1163
